File: kotlin-app/composeApp/src/commonMain/kotlin/app/madar/ui/MadarSheet.kt
package app.madar.ui

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.gestures.detectVerticalDragGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.roundToInt

// Custom modal bottom sheet — the Compose mirror of swift-app Components/
// MadarSheet.swift. The Kotlin app previously hand-rolled this per screen
// (ItemDetailSheet) with no enter/exit animation, and used Material AlertDialog
// elsewhere (Waiter settle), so modals diverged hard from Swift. This is the one
// shared presenter: a scrim + bottom-anchored branded card that slides in on a
// spring, dims the scrim, and supports tap-scrim + drag-down dismissal — then
// animates OUT before unmounting (the piece the old instant-pop sheets missed).
//
// Mount it conditionally like the Swift `.madarSheet(isPresented:)`:
//
//   if (model.showItemSheet) {
//       MadarSheet(onDismiss = { model.showItemSheet = false }) { dismiss ->
//           ItemDetailContent(..., onClose = dismiss)   // close button calls dismiss()
//       }
//   }
//
// `onDismiss` fires AFTER the slide-out completes, so the caller's boolean only
// flips once the card is off-screen — no hard cut.

/** How tall the card may grow (mirrors Swift `SheetSize`). */
enum class SheetSize {
    /** Fills ~88% of the container — the default for medium sheets. */
    AUTO,
    /** Reaches ~94% — big sheets (checkout / tender). */
    LARGE,
    /** Hugs its content up to a 92% cap and scrolls only on overflow — item /
     *  bundle customize sheets that must not stretch into a tall empty void. */
    HUG,
}

@Composable
fun MadarSheet(
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
    size: SheetSize = SheetSize.AUTO,
    maxWidth: Dp = Responsive.sheetMaxWidth,
    content: @Composable ColumnScope.(dismiss: () -> Unit) -> Unit,
) {
    val c = madarColors()
    val density = LocalDensity.current
    val scope = rememberCoroutineScope()

    // Drives the slide + scrim. `shown` is set true once, on first composition, so
    // animateFloatAsState animates the card up from below and fades the scrim in.
    var shown by remember { mutableStateOf(false) }
    LaunchedEffect(Unit) { shown = true }

    val shape = RoundedCornerShape(topStart = Radii.xl, topEnd = Radii.xl)

    BoxWithConstraints(modifier.fillMaxSize()) {
        val maxH: Dp = maxHeight * when (size) {
            SheetSize.LARGE -> 0.94f
            SheetSize.HUG -> 0.92f
            SheetSize.AUTO -> 0.88f
        }
        // Fully-hidden translation: the whole container height plus a margin so the
        // shadow clears the bottom edge too.
        val hiddenPx = with(density) { maxHeight.toPx() } + 80f

        // Base slide offset (spring) + the user's live drag offset (separate
        // Animatable so a released-but-not-dismissed drag snaps back smoothly).
        val baseOffset by animateFloatAsState(
            targetValue = if (shown) 0f else hiddenPx,
            animationSpec = MotionSpec.sheet(),
            label = "sheetOffset",
        )
        val dragY = remember { Animatable(0f) }
        val scrimAlpha by animateFloatAsState(
            targetValue = if (shown) Opacity.scrim else 0f,
            animationSpec = MotionSpec.standard(),
            label = "sheetScrim",
        )

        // Animate out, THEN clear the caller's binding (mirrors Swift `close()`).
        fun dismiss() {
            if (!shown) return
            shown = false
            scope.launch { delay(280); onDismiss() }
        }

        // Scrim — tap to dismiss. Sits below the card.
        Box(
            Modifier.fillMaxSize()
                .background(Color.Black.copy(alpha = scrimAlpha))
                .clickable(
                    interactionSource = remember { MutableInteractionSource() },
                    indication = null,
                ) { dismiss() },
        )

        // The card.
        Column(
            Modifier
                .align(Alignment.BottomCenter)
                .widthIn(max = maxWidth)
                .fillMaxWidth()
                .offset { IntOffset(0, (baseOffset + dragY.value).roundToInt()) }
                .then(if (size == SheetSize.HUG) Modifier.heightIn(max = maxH) else Modifier.height(maxH))
                .elevation(Elevation.RAISED, shape)
                .clip(shape)
                .background(c.surface)
                .border(1.dp, c.borderLight, shape),
        ) {
            // Grab handle — drag down past a threshold (or fling) to dismiss.
            Box(
                Modifier.fillMaxWidth()
                    .pointerInput(Unit) {
                        detectVerticalDragGestures(
                            onVerticalDrag = { _, delta ->
                                scope.launch { dragY.snapTo((dragY.value + delta).coerceAtLeast(0f)) }
                            },
                            onDragEnd = {
                                if (dragY.value > maxH.toPx() * 0.28f) dismiss()
                                else scope.launch { dragY.animateTo(0f, MotionSpec.sheet()) }
                            },
                            onDragCancel = {
                                scope.launch { dragY.animateTo(0f, MotionSpec.sheet()) }
                            },
                        )
                    },
                contentAlignment = Alignment.Center,
            ) {
                Box(
                    Modifier.padding(top = Space.md, bottom = Space.sm)
                        .size(width = 40.dp, height = 5.dp)
                        .clip(CircleShape)
                        .background(c.border),
                )
            }

            content(::dismiss)
        }
    }
}

File: swift-app/Sources/MadarUI/Components/MadarSheet.swift
// Custom modal bottom-sheet presentation.
//
// SwiftUI's `.sheet` on macOS is a rigid modal panel: no drag-to-dismiss, no
// tap-outside-to-dismiss, and stacking many on one node leaves dismiss artifacts
// (the "weird artifacting after the orders popup"). This presents a scrim + a
// bottom-anchored card entirely within the view tree, so it:
//   • inherits the environment (theme/localize/RTL/toast) — no modalChrome needed,
//   • supports tap-scrim + drag-down dismissal like Flutter's showModalBottomSheet,
//   • renders identically on macOS and iOS.
//
// Use `.madarSheet(isPresented:)` / `.madarSheet(item:)` exactly like `.sheet`.
import SwiftUI

/// How tall the card may grow. `.auto` fits its content up to a cap; `.large`
/// reaches ~94% of the container (for big sheets like checkout); `.hug` sizes
/// strictly to its content (capped at 92%) and scrolls only when it overflows —
/// the item/bundle customize sheets, which must not stretch to a tall empty void
/// for a simple item (mirrors Flutter's `MainAxisSize.min` bottom sheet).
enum SheetSize { case auto, large, hug }

extension View {
    /// Present a custom bottom sheet driven by a boolean. The content builder gets
    /// an animated `dismiss` closure — wire it to the content's close button so the
    /// card slides out (not a hard cut).
    func madarSheet<SheetBody: View>(
        isPresented: Binding<Bool>,
        size: SheetSize = .auto,
        maxWidth: CGFloat = 600,
        onDismiss: (() -> Void)? = nil,
        @ViewBuilder content: @escaping (_ dismiss: @escaping () -> Void) -> SheetBody
    ) -> some View {
        overlay {
            if isPresented.wrappedValue {
                MadarSheetContainer(size: size, maxWidth: maxWidth, requestClose: {
                    isPresented.wrappedValue = false
                    onDismiss?()
                }) { dismiss in content(dismiss) }
            }
        }
    }

    /// Present a custom bottom sheet driven by an optional identifiable item.
    func madarSheet<Item: Identifiable, SheetBody: View>(
        item: Binding<Item?>,
        size: SheetSize = .auto,
        maxWidth: CGFloat = 600,
        onDismiss: (() -> Void)? = nil,
        @ViewBuilder content: @escaping (_ item: Item, _ dismiss: @escaping () -> Void) -> SheetBody
    ) -> some View {
        overlay {
            if let value = item.wrappedValue {
                MadarSheetContainer(size: size, maxWidth: maxWidth, requestClose: {
                    item.wrappedValue = nil
                    onDismiss?()
                }) { dismiss in content(value, dismiss) }
            }
        }
    }
}

extension View {
    /// Present a full-screen ROUTED screen (not a sheet) that slides in from the
    /// trailing edge over the order hub — the macOS-safe equivalent of a navigation
    /// push (macOS has no `fullScreenCover`). The content fills the window and
    /// supplies its own back-chevron header wired to `dismiss`. Keeps the design
    /// rule that the order action bar is the only nav hub: the bar pushes screens.
    func appScreen<ScreenBody: View>(
        isPresented: Binding<Bool>,
        @ViewBuilder content: @escaping (_ dismiss: @escaping () -> Void) -> ScreenBody
    ) -> some View {
        modifier(AppScreenModifier(isPresented: isPresented, screen: content))
    }
}

private struct AppScreenModifier<ScreenBody: View>: ViewModifier {
    @Environment(\.theme) private var theme
    @Binding var isPresented: Bool
    @ViewBuilder var screen: (_ dismiss: @escaping () -> Void) -> ScreenBody

    func body(content: Content) -> some View {
        content.overlay {
            ZStack {
                if isPresented {
                    screen { withAnimation(Motion.standard) { isPresented = false } }
                        .frame(maxWidth: .infinity, maxHeight: .infinity)
                        .background(theme.colors.bg.ignoresSafeArea())
                        .transition(.move(edge: .trailing))
                        .zIndex(20)
                }
            }
            .animation(Motion.standard, value: isPresented)
        }
    }
}

/// The scrim + sliding card. Owns its enter/exit animation so user dismissals
/// (tap-scrim, drag-down, close button) animate out before the binding clears.
private struct MadarSheetContainer<SheetBody: View>: View {
    @Environment(\.theme) private var theme
    let size: SheetSize
    /// Flutter caps every bottom sheet at `ResponsiveSheet` = 600; that's the
    /// default here too. Narrower, focused sheets (item/bundle customize) pass a
    /// smaller value (~540) at the call site.
    var maxWidth: CGFloat = 600
    let requestClose: () -> Void
    @ViewBuilder var sheetBody: (_ dismiss: @escaping () -> Void) -> SheetBody

    @State private var shown = false
    @State private var dragY: CGFloat = 0

    private let spring = Animation.spring(response: 0.34, dampingFraction: 0.9)

    /// Animate the card out, then clear the caller's binding.
    private func close() {
        withAnimation(spring) { shown = false; dragY = 0 }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.30) { requestClose() }
    }

    var body: some View {
        GeometryReader { geo in
            let maxH = geo.size.height * (size == .large ? 0.94 : size == .hug ? 0.92 : 0.88)
            ZStack(alignment: .bottom) {
                // Scrim — tap to dismiss.
                Color.black.opacity(shown ? 0.45 : 0)
                    .ignoresSafeArea()
                    .contentShape(Rectangle())
                    .onTapGesture { close() }

                // Card. `.hug` lets the body size to its own content (capped at
                // maxH); other sizes force-fill the available height.
                VStack(spacing: 0) {
                    grabber
                    if size == .hug {
                        // Fill the card width (600), but hug height (no maxHeight).
                        sheetBody { close() }
                            .frame(maxWidth: .infinity)
                    } else {
                        sheetBody { close() }
                            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
                    }
                }
                // Constrain the card to `maxWidth` and apply ALL its chrome
                // (background, clip, border, shadow) to THAT constrained frame —
                // BEFORE the full-width centering frame. The old order applied the
                // background after `.frame(maxWidth: .infinity)`, so the card's
                // background+border stretched edge-to-edge (the "takes the whole
                // width" bug) even though the content sat in a 600pt column.
                .frame(maxWidth: maxWidth)
                .frame(maxHeight: maxH, alignment: .bottom)
                .background(theme.colors.surface)
                .clipShape(TopRoundedRectangle(radius: Radii.xl))
                .overlay(
                    TopRoundedRectangle(radius: Radii.xl)
                        .strokeBorder(theme.colors.borderLight, lineWidth: 1)
                )
                .shadow(color: theme.colors.shadow.opacity(shown ? 1 : 0), radius: 30, y: -8)
                .frame(maxWidth: .infinity) // center the finished card on wide windows
                .offset(y: shown ? dragY : maxH + 60)
                .gesture(dragToDismiss(maxH: maxH))
                .ignoresSafeArea(.container, edges: .bottom)
            }
        }
        .onAppear { withAnimation(spring) { shown = true } }
        #if os(iOS)
        .ignoresSafeArea(.keyboard)
        #endif
    }

    private var grabber: some View {
        Capsule()
            .fill(theme.colors.border)
            .frame(width: 40, height: 5)
            .padding(.top, Space.md)
            .padding(.bottom, Space.sm)
            .frame(maxWidth: .infinity)
            .background(theme.colors.surface) // same surface as the sheet — continuous top
            .contentShape(Rectangle())
            .gesture(dragToDismiss(maxH: 600))
    }

    /// Drag the card down; release past a threshold (or with downward flick) closes.
    private func dragToDismiss(maxH: CGFloat) -> some Gesture {
        DragGesture(minimumDistance: 6)
            .onChanged { v in
                if v.translation.height > 0 { dragY = v.translation.height }
            }
            .onEnded { v in
                let flung = v.predictedEndTranslation.height > 220
                if dragY > maxH * 0.28 || flung {
                    close()
                } else {
                    withAnimation(spring) { dragY = 0 }
                }
            }
    }
}

/// A rectangle with only its top two corners rounded — version-safe (no
/// UnevenRoundedRectangle, which needs macOS 13.3 / iOS 16.4).
struct TopRoundedRectangle: InsettableShape {
    var radius: CGFloat
    var inset: CGFloat = 0

    func path(in rect: CGRect) -> Path {
        let r = min(radius, min(rect.width, rect.height) / 2)
        let rr = rect.insetBy(dx: inset, dy: inset)
        var p = Path()
        p.move(to: CGPoint(x: rr.minX, y: rr.maxY))
        p.addLine(to: CGPoint(x: rr.minX, y: rr.minY + r))
        p.addArc(center: CGPoint(x: rr.minX + r, y: rr.minY + r),
                 radius: r, startAngle: .degrees(180), endAngle: .degrees(270), clockwise: false)
        p.addLine(to: CGPoint(x: rr.maxX - r, y: rr.minY))
        p.addArc(center: CGPoint(x: rr.maxX - r, y: rr.minY + r),
                 radius: r, startAngle: .degrees(270), endAngle: .degrees(0), clockwise: false)
        p.addLine(to: CGPoint(x: rr.maxX, y: rr.maxY))
        p.closeSubpath()
        return p
    }

    func inset(by amount: CGFloat) -> TopRoundedRectangle {
        TopRoundedRectangle(radius: radius, inset: inset + amount)
    }
}
